packages/client/src/runtime/DataProxyEngine.ts

```diff
@@ -0,0 +1,243 @@
+  private headers: { Authorization: string }
+
+  private logEmitter = new EventEmitter()
+
+  constructor(private config: EngineConfig) {
+    super()
+
+    this.schemaText = fs.readFileSync(config.datamodelPath, 'utf8')
+    this.schemaBase64 = Buffer.from(this.schemaText).toString('base64')
+    this.schemaHash = createHash('sha256')
+      .update(this.schemaBase64)
+      .digest('hex')
+
+    const [host, apiKey] = extractHostAndApiKey(this.schemaText)
+    // TODO: how to get the current version the right way?
+    const clientVersion = '2.29.1'
+
+    this.url = (s) => `https://${host}/${clientVersion}/${this.schemaHash}/${s}`
+    this.headers = { Authorization: `Bearer ${apiKey}` }
+
+    this.logEmitter.on('error', () => {
+      // Prevent unhandled error events
+    })
+
+    this.logEmitter.emit('info', `Ready for ${host}`)
+  }
+
+  version() {
+    // QE is remote, we don't need to know the exact commit SHA
+    return 'unknown'
+  }
+
+  async start() {}
+  async stop() {}
+
+  on(event: EngineEventType, listener: (args?: any) => any): void {
+    if (event === 'beforeExit') {
+      // TODO: hook into the process
+      throw new Error('beforeExit event is not supported yet')
+    } else {
+      this.logEmitter.on(event, listener)
+    }
+  }
+
+  // TODO: looks like activeProvider is the only thing
+  // used externally; verify that
+  getConfig() {
+    return Promise.resolve({
+      datasources: [
+        {
+          activeProvider: this.config.activeProvider,
+        },
+      ],
+    } as GetConfigResult)
+  }
+
+  private async uploadSchema() {
+    const res = await fetch(this.url('schema'), {
+      method: 'PUT',
+      headers: this.headers,
+      body: this.schemaBase64,
+    })
+
+    if (!res.ok) {
+      this.logEmitter.emit('warn', 'Could not upload the schema')
+      throw new Error('Could not upload the schema')
+    }
+  }
+
+  async request<T>(q: string, _hs?: any, n = 0) {
+    try {
+      this.logEmitter.emit('query', q)
+
+      const res = await fetch(this.url('graphql'), {
+        method: 'POST',
+        headers: this.headers,
+        body: JSON.stringify({
+          operationName: null,
+          variables: {},
+          query: q,
+        }),
+      })
+
+      // 404 on the GraphQL endpoint may mean that the schema
+      // was not uploaded yet.
+      if (res.status === 404) {
+        await this.uploadSchema()
+        throw new Error('Schema (re)uploaded')
+      }
+
+      if (!res.ok) {
+        throw new Error('GraphQL request failed')
+      }
+
+      return JSON.parse(await res.text())
+    } catch (err) {
+      if (n >= MAX_RETRIES) {
+        this.logEmitter.emit('error', `Failed to query: ${err.message}`)
+        throw err
+      } else {
+        const delay = await backOff(n)
+        this.logEmitter.emit('warn', `Retrying after ${delay}`)
+        return this.request<T>(q, {}, n + 1)
+      }
+    }
+  }
+
+  requestBatch<T>(qs: string[], _hs: any, transact = false) {
```

Comment:
Would it be possible to rename the arguments in the code? 
I guess
- `qs` is `querystring` but it's a `string[]` and I also see `q` as `string` so renaming could help maybe? making it more explicit what this is.
- `_hs` is `headers`
- `transact` is `transaction`/`runInTransaction`